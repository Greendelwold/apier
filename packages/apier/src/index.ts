import { ApierKind, kindOf } from '@dee-contrib/apier-utils';
import { ApierComment } from '@dee-contrib/apier-comment';

export abstract class ApierItem {
  public readonly comment: ApierComment;
  public readonly name: string;
  public readonly value: any;
  public readonly model: any;
  constructor(comment: ApierComment, name: string, value: any) {
    this.comment = comment.scope([name]);
    this.name = name;
    this.value = value;
  }
  public kind(): ApierKind {
    return kindOf(this.value);
  }
  protected createJSONKind(comment: ApierComment, name: string, value: any): ApierJSONKind {
    switch (kindOf(value)) {
      case ApierKind.INTEGER:
        return new ApierInteger(comment, name, value);
      case ApierKind.NUMBER:
        return new ApierNumber(comment, name, value);
      case ApierKind.STRING:
        return new ApierString(comment, name, value);
      case ApierKind.NULL:
        return new ApierNull(comment, name, value);
      case ApierKind.BOOLEAN:
        return new ApierBoolean(comment, name, value);
      case ApierKind.ARRAY:
        return new ApierArray(comment, name, value);
      case ApierKind.OBJECT:
        return new ApierObject(comment, name, value);
      default:
        throw new Error('unreachable');
    }
  };
}

export interface ApierRawObject {
  [k: string]: any;
}

export type ApierJSONKind = ApierArray | ApierObject | ApierParameter
export type ApierParameter = ApierInteger | ApierString | ApierBoolean | ApierNull | ApierNumber;
export interface ApierReqModel {
  params?: ApierObject;
  query?: ApierObject;
  headers?: ApierObject;
  body?: ApierJSONKind;
}
export class ApierReq extends ApierItem {
  public readonly model: ApierReqModel;
  public readonly value: ApierRawReq;
  constructor(comment: ApierComment, name: string, value: ApierRawReq) {
    super(comment, name, value);
    this.model = Object.keys(value).reduce((model, key) => {
      model[key] = this.createJSONKind(this.comment, key, value[key]);
      return model;
    }, {});
  }
}
export interface ApierResModel {
  status: number;
  body?: ApierJSONKind;
}
export class ApierRes extends ApierItem {
  public readonly model: ApierResModel;
  public readonly value: ApierRawRes;
  constructor(comment: ApierComment, name: string, value: ApierRawRes) {
    super(comment, name, value);
    const { status, body } = value;
    this.model = { status };
    if (body) this.model.body = this.createJSONKind(this.comment, 'body', body);
  }
}
export class ApierNumber extends ApierItem {
  public readonly model = null;
  public readonly value: number;
  constructor(comment: ApierComment, name: string, value: number) {
    super(comment, name, value);
  }
}

export class ApierInteger extends ApierItem {
  public readonly model = null;
  public readonly value: number;
  constructor(comment: ApierComment, name: string, value: number) {
    super(comment, name, value);
  }
}
export class ApierString extends ApierItem {
  public readonly model = null;
  public readonly value: string;
  constructor(comment: ApierComment, name: string, value: string) {
    super(comment, name, value);
  }
}
export class ApierBoolean extends ApierItem {
  public readonly model = null;
  public readonly value: boolean;
  constructor(comment: ApierComment, name: string, value: boolean) {
    super(comment, name, value);
  }
}
export class ApierNull extends ApierItem {
  public readonly model = null;
  public readonly value: null;
  constructor(comment: ApierComment, name: string, value: null) {
    super(comment, name, value);
  }
}
export type ApierArrayModel = ApierJSONKind[];
export class ApierArray extends ApierItem {
  public readonly model: ApierArrayModel;
  public readonly value: any[];
  constructor(comment: ApierComment, name: string, value: any[]) {
    super(comment, name, value);
    this.model = value.map((v, i) => this.createJSONKind(this.comment, String(i), v));
  }
}
export type ApierObjectModel = {
  [k: string]: ApierJSONKind;
}

export class ApierObject extends ApierItem {
  public readonly model: ApierObjectModel;
  public readonly value: ApierRawObject;
  constructor(comment: ApierComment, name: string, value: ApierRawObject) {
    super(comment, name, value);
    this.model = Object.keys(value).reduce((model, key) => {
      model[key] = this.createJSONKind(this.comment, key, value[key]);
      return model;
    }, {});
  }
}

export enum Method {
  GET = 'get',
  POST = 'post',
  PUT = 'put',
  DELETE = 'delete',
}

export interface ApierRaw {
  name: string;
  url: string;
  method: Method;
  req?: ApierRawReq;
  res: ApierRawRes;
}

export interface ApierRawReq {
  params?: ApierRawParameters;
  query?: ApierRawParameters;
  headers?: ApierRawParameters;
  body?: any;
}

export interface ApierRawParameters {
  [k: string]: number | string | null | boolean;
}

export interface ApierRawRes {
  status?: number;
  body?: any;
}

export interface ParseResult {
  apis: {
    [k: string]: ApierRaw;
  }
  comment: ApierComment;
}

export interface Parser {
  parse(input: string): ParseResult;
}

export class ParserError extends Error {
  public readonly paths: string[];
  constructor(paths: string[], message: string) {
    super(message);
    this.paths = paths;
  }
}

export interface ApierModel {
  req?: ApierReq;
  res: ApierRes;
}
export class Apier extends ApierItem {
  public readonly method: Method;
  public readonly url: string;
  public readonly model: ApierModel
  public readonly value: ApierRaw;
  constructor(comment: ApierComment, value: ApierRaw) {
    super(comment, value.name, value);
    this.method = value.method;
    this.url = value.url;
    const model: any = {};
    if (value.req) model.req = new ApierReq(this.comment, 'req', value.req);
    model.res = new ApierRes(this.comment, 'res', value.res);
    this.model = model;
  }
}

export function parse(input: string, parser: Parser): Apier[] {
  const result = [];
  const { apis, comment } = parser.parse(input);
  for (const name in apis) {
    result.push(new Apier(comment, apis[name]));
  }
  return result;
}