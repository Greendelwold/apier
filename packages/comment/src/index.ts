import omit from "lodash/omit";
import pick from "lodash/pick";

// 注释键值, `optional`, `type=integer`, `description="split with ws"`
const RG_COMMENT_VALUE = /([A-Za-z_\-]+)(\=("[^"]*"|'[^']*'|[^\s]+))?/g;

export interface CommentObject {
  [k: string]: any;
}
export interface CommentItem {
  paths: string[];
  comment: CommentObject;
}

export class ApierComment {
  public readonly comments: CommentItem[];
  constructor(comments = []) {
    this.comments = comments;
  }
  public append(paths: string[], commentText: string) {
    const comment = this.parse(commentText);
    if (comment) this.comments.push({ paths, comment });
  }
  public scope(paths: string[]) {
    const comments = this.comments
      .filter(c => isPrefixArray(paths, c.paths))
      .map(c => ({ paths: c.paths.slice(paths.length), comment: c.comment }));
    return new ApierComment(comments);
  }
  public retrive(paths: string[] = []): CommentUtil {
    const commentItem = this.comments.find(c => paths.length === c.paths.length && isPrefixArray(paths, c.paths));
    return new CommentUtil(commentItem ? commentItem.comment : {});
  }
  /**
   * 获取行注释
   *
   * `optional type=integer format=int32` => { optional: true, type: integer, format: int32 }
   */
  private parse(commentText: string): { [k: string]: any } {
    const result = {};
    let matched;
    while ((matched = RG_COMMENT_VALUE.exec(commentText))) {
      const key = matched[1];
      let value = matched[3];
      if (value === undefined) {
        value = true;
      } else {
        try {
          value = JSON.parse(value);
        } catch (err) {}
      }
      result[key] = value;
    }
    return result;
  }
}

export class CommentUtil {
  private readonly comment: CommentObject;
  constructor(comment: CommentObject) {
    this.comment = comment;
  }
  omit(keys: string[]) {
    return omit(this.comment, keys);
  }
  pick(keys: string[]) {
    return pick(this.comment, keys);
  }
  modify(key: string, value: any) {
    this.comment[key] = value;
  }
  val(key?: string, defaultValue?: any) {
    if (key === undefined) return { ...this.comment };
    if (this.comment.hasOwnProperty(key)) {
      return this.comment[key];
    }
    return defaultValue;
  }
}

function isPrefixArray<T>(prefixArr: T[], arr: T[]) {
  return prefixArr.every((v, i) => arr[i] === v);
}
