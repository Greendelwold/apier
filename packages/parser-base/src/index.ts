import { ApierComment } from "@jiasuyun/apier-comment";

export interface ApierRaw {
  name: string;
  url: string;
  method: Method;
  req?: ApierRawReq;
  res: ApierRawRes;
}

export interface ApierRawObject {
  [k: string]: any;
}

export interface ApierRawReq {
  params?: ApierRawParameters;
  query?: ApierRawParameters;
  headers?: ApierRawParameters;
  body?: any;
}

export interface ApierRawParameters {
  [k: string]: number | string | null | boolean;
}

export interface ApierRawRes {
  status?: number;
  body?: any;
}

export interface ParseResult {
  apis: {
    [k: string]: ApierRaw;
  };
  comment: ApierComment;
}

export interface Parser {
  parse(input: string): ParseResult;
}

export class ParserError extends Error {
  public readonly paths: string[];
  constructor(paths: string[], message: string) {
    super(message);
    this.paths = paths;
  }
}

export enum Method {
  GET = "get",
  POST = "post",
  PUT = "put",
  DELETE = "delete"
}
