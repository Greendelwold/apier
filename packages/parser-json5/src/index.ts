import * as apier from "@jiasuyun/apier-parser-base";
import { ApierComment } from "@jiasuyun/apier-comment";
import { ApierKind, kindOf } from "@jiasuyun/apier-utils";
import * as JSON5 from "json5";
import Visitor from "./Visitor";
import { beignLineNum } from "./helper";
import lset from "lodash/set";

// 解析 Route
const RE_ROUTE = /^(get|post|put|delete)\s[:\/A-Za-z0-9_\-]+/i;
// 判断根注释, `// @@@`
const RE_ROOT_COMMENT = /^\s*\/\/\s*@@@/;

export default class Parser implements apier.Parser {
  public parse(input: string): apier.ParseResult {
    let parsedObj: any;
    try {
      parsedObj = JSON5.parse(input);
    } catch (err) {
      throw new apier.ContentParserError(err.lineNumber, err.columnNumber, err);
    }
    const apis = {};
    for (const name in parsedObj) {
      apis[name] = this.parseApi(name, parsedObj[name]);
    }
    const comment = this.parseComment(input);
    return { apis, comment };
  }
  private parseApi(name: string, data: any): apier.ApierRaw {
    if (kindOf(data) !== ApierKind.OBJECT) {
      throw new apier.StructParserError([name], "should be object");
    }
    let api: any = { name };
    const { route, req, res } = data;
    this.parseRoute(api, route);
    this.parseReq(api, req);
    this.parseRes(api, res);
    return api as apier.ApierRaw;
  }
  private parseComment(input: string): ApierComment {
    const lines = input.split("\n");
    const comment = new ApierComment();
    const root = new Visitor(lines, comment, []);
    const beginLineIndex = beignLineNum(lines);
    this.parserRootComment(comment, lines.slice(0, beginLineIndex));
    root.scopeObject(beginLineIndex + 1);
    return comment;
  }
  private parserRootComment(comment: ApierComment, lines: string[]) {
    const paths = [];
    for (let line of lines) {
      const match = RE_ROOT_COMMENT.exec(line);
      if (!match) {
        continue;
      }
      comment.append(paths, line.slice(match[0].length));
    }
  }
  private parseRoute(api: apier.ApierRaw, route: string) {
    if (!RE_ROUTE.test(route)) {
      throw new apier.StructParserError([api.name, "route"], route);
    }
    const [method, url] = route.split(" ");
    api.method = method.toLowerCase() as apier.Method;
    api.url = url;
  }

  private parseReq(api: apier.ApierRaw, req) {
    if (req === undefined) return;
    if (kindOf(req) !== ApierKind.OBJECT) {
      throw new apier.StructParserError([api.name, "req"], "must be object");
    }
    const { headers, params, query, body } = req;
    api.req = {};
    if (headers) this.parseParameters(api, ["req", "headers"], headers);
    if (params) this.parseParameters(api, ["req", "params"], params);
    if (query) this.parseParameters(api, ["req", "query"], query);
    if (body) this.parseBody(api, ["req", "body"], body);
  }

  private parseRes(api: apier.ApierRaw, res) {
    api.res = { status: 200 };
    if (res === undefined) {
      return;
    }
    if (kindOf(res) !== ApierKind.OBJECT) {
      throw new apier.StructParserError([api.name, "res"], "must be object");
    }
    const { status, body } = res;
    if (status) this.parseResStatus(api, status);
    if (body) this.parseBody(api, ["res", "body"], body);
  }

  private parseParameters(api, paths, obj) {
    if (typeof obj !== "object") {
      throw new apier.StructParserError([api.name, ...paths], `must be object`);
    }
    for (const key in obj) {
      const value = obj[key];
      const valueKind = kindOf(value);
      if (valueKind === ApierKind.ARRAY || valueKind === ApierKind.OBJECT) {
        throw new apier.StructParserError([api.name, ...paths, key], `must be scalar value`);
      }
    }
    lset(api, paths, obj);
  }

  private parseResStatus(api, status = 200) {
    if (typeof status !== "number" && (status < 100 && status >= 600)) {
      throw new apier.StructParserError([name, "res", "status"], `{status}`);
    }
    api.res.status = status;
  }

  private parseBody(api, paths, body) {
    lset(api, paths, body);
  }
}
