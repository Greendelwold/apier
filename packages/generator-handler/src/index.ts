import * as apier from "@jiasuyun/apier"; // eslint-disable-line

export type GeneratorResult = string;
export default class Generator {
  public readonly value: GeneratorResult;
  constructor(apier: apier.Apier) {
    this.value = `export async function ${
      apier.name
    }(req: Request, res: Response) {}`;
  }
}
