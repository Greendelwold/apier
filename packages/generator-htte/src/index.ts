import { colonToCurlybrace } from '@dee-contrib/apier-utils';
import * as apier from '@dee-contrib/apier';

export interface GeneratorResult {
  describe: string;
  req: {
    method: string;
    url: string;
  } & apier.ApierRawReq;
  res: apier.ApierRawRes;
}

export class Generator {
  public readonly value: GeneratorResult;
  constructor(apier: apier.Apier) {
    const { method, url, name, value: { req, res } } = apier;
    const summary = apier.comment.retrive().val('summary', name) as string;
    this.value = {
      describe: summary,
      req: { method, url: colonToCurlybrace(url), ...req },
      res
    }
  }
}