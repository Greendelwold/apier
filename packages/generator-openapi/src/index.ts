import * as apier from "@jiasuyun/apier";
import * as openapi from "openapi3-ts";
import { ApierKind, colonToCurlybrace } from "@jiasuyun/apier-utils";
import { OPERATION_KEYS, PARAMETER_KEYS, SCHEMA_KEYS } from "./constants";
export interface GeneratorResult {
  paths: openapi.PathsObject;
  components: openapi.ComponentsObject;
}

export default class Generator {
  public readonly value: GeneratorResult;
  private apier: apier.Apier;
  private operation: openapi.OperationObject;
  constructor(apier: apier.Apier) {
    const schemas: openapi.SchemasObject = {};
    const operation = {};
    const url = colonToCurlybrace(apier.url);
    this.apier = apier;
    this.value = {
      paths: { [url]: { [apier.method]: operation } },
      components: { schemas }
    };
    this.operation = operation as openapi.OperationObject;
    this.generate();
  }

  generate() {
    const { apier } = this;
    const {
      comment,
      model: { req }
    } = apier;
    const commentUtil = comment.retrive();
    const operation = this.operation;
    operation.operationId = apier.name;
    const summary = commentUtil.val("summary", apier.name);
    operation.summary = summary;
    const parameters = this.dealParameters();
    if (parameters.length > 0) {
      operation.parameters = parameters;
    }
    if (req.model.body) this.dealReqeustBody();
    this.dealResponses();
    Object.assign(operation, commentUtil.pick(OPERATION_KEYS));
  }
  dealParameters(): (openapi.ParameterObject | openapi.ReferenceObject)[] {
    const parameters = [];
    ["params", "query", "headers"].forEach(key => {
      const apierParameters: apier.ApierObject = this.apier.model.req.model[key];
      if (!apierParameters) return;
      Object.keys(apierParameters.model).forEach(name => {
        const apierParameter: apier.ApierItem = apierParameters.model[name];
        const commentUtil = apierParameter.comment.retrive();
        const useSchema = commentUtil.val("useSchema");
        if (useSchema) return parameters.push(createRef(useSchema));
        const parameter: openapi.ParameterObject = {
          name,
          in: inOfParameter(key)
        };
        Object.assign(parameter, commentUtil.pick(PARAMETER_KEYS));
        parameter.schema = this.createSchema(apierParameter, { isParameter: true });
        PARAMETER_KEYS.forEach(key => delete parameter.schema[key]);
        if (!commentUtil.val("optional")) parameter.required = true;
        const saveSchema = commentUtil.val("saveSchema");
        if (saveSchema) {
          parameters.push(this.saveParamters(saveSchema, parameter));
        } else {
          parameters.push(parameter);
        }
      });
    });
    return parameters;
  }
  dealReqeustBody() {
    const body: apier.ApierJSONKind = this.apier.model.req.model.body;
    const commentUtil = body.comment.retrive();
    const requestBody: openapi.RequestBodyObject = {} as any;
    this.operation.requestBody = requestBody;
    const description = commentUtil.val("description");
    if (description) requestBody.description = description;
    if (!commentUtil.val("optional")) requestBody.required = true;
    const useSchema = commentUtil.val("useSchema");
    const contentType = commentUtil.val("contentType", "application/json");
    if (useSchema)
      return (requestBody.content = {
        [contentType]: { schema: createRef(useSchema) }
      });
    if (!commentUtil.val("saveSchema")) {
      commentUtil.modify("saveSchema", nameOfReqResSchema(this.apier.name, "Request"));
    }
    requestBody.content = {
      [contentType]: {
        schema: this.createSchema(body)
      }
    };
  }
  dealResponses() {
    const res: apier.ApierRes = this.apier.model.res;
    const commentUtil = res.comment.retrive();
    const { status, body } = res.model;
    const responses: openapi.ResponsesObject = { [status]: {} };
    this.operation.responses = responses;
    responses[status].description = commentUtil.val("description", "SUCCESS");
    const contentType = commentUtil.val("contentType", "application/json");
    const useSchema = commentUtil.val("useSchema");
    if (useSchema)
      return (responses[status].content = {
        [contentType]: { schema: createRef(useSchema) }
      });
    if (!commentUtil.val("saveSchema")) {
      commentUtil.modify("saveSchema", nameOfReqResSchema(this.apier.name, "Request"));
    }
    responses[status].content = {
      [contentType]: {
        schema: body ? this.createSchema(body) : this.saveSchema(commentUtil.val("saveSchema"), { type: "object " })
      }
    };
  }
  saveSchema(name: string, schema: openapi.SchemaObject): openapi.ReferenceObject {
    this.value.components.schemas[name] = schema;
    return createRef(name);
  }
  saveParamters(name: string, schema: openapi.ParameterObject): openapi.ReferenceObject {
    let parameters = this.value.components.parameters;
    if (!parameters) parameters = this.value.components.parameters = {};
    parameters[name] = schema;
    return createRef(name, "parameters");
  }
  createSchema(
    apierItem: apier.ApierJSONKind,
    options: CreateSchemaOptions = { isParameter: false }
  ): openapi.SchemaObject {
    const schema: openapi.SchemaObject = {};
    this.schemaUtil(apierItem, { schema, ...options });
    return schema;
  }
  schemaUtil(apierItem: apier.ApierItem, context: SchemaUtilContext): boolean {
    const { schema, isParameter } = context;
    const commentUtil = apierItem.comment.retrive();
    const useSchema = commentUtil.val("useSchema");
    if (!isParameter && useSchema) {
      Object.assign(schema, createRef(useSchema));
      return !commentUtil.val("optional", false);
    }
    Object.assign(schema, { type: apierItem.kind() });
    if (apierItem instanceof apier.ApierObject) {
      this.schemaUtilObject(apierItem, context);
    } else if (apierItem instanceof apier.ApierArray) {
      this.schemaUtilArray(apierItem, context);
    }
    Object.assign(schema, commentUtil.pick(SCHEMA_KEYS));
    const saveSchema = commentUtil.val("saveSchema");
    if (!isParameter && saveSchema) {
      this.value.components.schemas[saveSchema] = { ...schema };
      Object.keys(schema).forEach(key => delete schema[key]); // clear
      Object.assign(schema, createRef(saveSchema));
    }
    return !commentUtil.val("optional", false);
  }
  schemaUtilArray(apierItem: apier.ApierArray, context: SchemaUtilContext) {
    const { schema } = context;
    const children = apierItem.model;
    if (children.length === 0) return;
    const commentUtil = apierItem.comment.retrive();
    let strategy = commentUtil.val("array", "anyOf");
    const items: any = (schema.items = {});
    if (children.length === 1) strategy = "first";
    if (strategy === "first") {
      this.schemaUtil(children[0], { ...context, schema: items });
    } else if (strategy === "anyOf") {
      const anyOf = (items.anyOf = []);
      children.forEach(child => {
        const childSchema = {};
        this.schemaUtil(child, { ...context, schema: childSchema });
        anyOf.push(childSchema);
      });
    } else {
      this.schemaUtilArrayAll(apierItem, context);
    }
  }
  schemaUtilArrayAll(apierItem: apier.ApierArray, context: SchemaUtilContext) {
    const { schema } = context;
    let requiredAll = [];
    let properties = {};
    const children = apierItem.model;
    children.forEach(child => {
      const childSchema: openapi.SchemaObject = {};
      this.schemaUtil(child, { ...context, schema: childSchema });
      if (child.kind() === ApierKind.OBJECT) {
        requiredAll = [...requiredAll, ...childSchema.required];
        properties = { ...properties, ...childSchema.properties };
      }
    });
    schema.type = "object";
    schema.properties = properties;
    const required = filterByCount(requiredAll, children.length);
    if (required.length > 0) schema.required = required;
  }
  schemaUtilObject(apierItem: apier.ApierObject, context: SchemaUtilContext) {
    const { schema } = context;
    const properties = {};
    const required = [];
    let haveProperties = false;
    for (const key in apierItem.model) {
      haveProperties = true;
      const item = (properties[key] = {});
      if (this.schemaUtil(apierItem.model[key], { ...context, schema: item })) {
        required.push(key);
      }
    }
    if (haveProperties) schema.properties = properties;
    if (required.length > 0) schema.required = required;
  }
}

function createRef(name: string, kind: string = "schemas"): openapi.ReferenceObject {
  return { ["$ref"]: `#/components/${kind}/${name}` };
}

function inOfParameter(name: string): openapi.ParameterLocation {
  if (name === "params") return "path";
  if (name === "query") return "query";
  return "header";
}

function nameOfReqResSchema(text, tail) {
  return [text[0].toUpperCase(), ...text.slice(1), ...tail].join("");
}

function filterByCount(arr: string[], count: number): string[] {
  const result = [];
  arr.sort().reduce(
    (a, c) => {
      if (c === a.prev) {
        a.count++;
        if (a.count === count) {
          result.push(c);
        }
        return a;
      } else {
        return { prev: c, count: 1 };
      }
    },
    { prev: "", count: 0 }
  );
  return result;
}

export interface CreateSchemaOptions {
  isParameter?: boolean;
}
export interface SchemaUtilContext extends CreateSchemaOptions {
  schema: openapi.SchemaObject;
}
