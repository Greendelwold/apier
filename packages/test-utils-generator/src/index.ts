import * as apier from "@jiasuyun/apier";
import Parser from "@jiasuyun/apier-parser-json5";
import { loadFixtureJSON5 } from "@jiasuyun/apier-test-utils";
import yaml from "js-yaml";
import merge from "lodash/merge";

export function parse(name: string): apier.ParseResult {
  const input = loadFixtureJSON5(name);
  const parser = new Parser();
  return apier.parse(input, parser);
}

export function toYaml(data: any): string {
  return yaml.safeDump(data);
}

export type GenerateFn = (apier: apier.Apier) => any;

export function factory(fn: GenerateFn) {
  return (name: string) => {
    const input = parse(name);
    return input.apiers.map(a => fn(a));
  };
}

export function mergeArray(arr: string[]) {
  return arr.reduce((a, c) => merge(a, c), {});
}
