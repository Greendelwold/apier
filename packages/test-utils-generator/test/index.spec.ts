import { parse, factory } from "../src";
import { loadFixtureJSON } from "@jiasuyun/apier-test-utils";

test("parse", () => {
  const result = parse("general");
  expect(result.apiers).toHaveLength(1);
  expect(result.apis).toEqual(loadFixtureJSON("general.apis"));
  const commentsJSON = loadFixtureJSON("general.comments");
  expect(result.comments).toEqual(commentsJSON);
  expect(result.metadata).toEqual(commentsJSON[0].comment);
});

test("generateYaml", () => {
  const fn = apier => ({ [apier.name]: true });
  expect(factory(fn)("user")).toMatchSnapshot();
});
