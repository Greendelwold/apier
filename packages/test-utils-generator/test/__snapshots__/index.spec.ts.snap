// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateYaml 1`] = `
"signup: true
loginPasswd: true
getGcode: true
loginSms: true
getScode: true
updatePasswd: true
listFavoriates: true
checkProductFavoriate: true
createFeedback: true
getStats: true
getUserInfo: true
updateUserInfo: true
getUserSettings: true
updateUserSettings: true
"
`;
